### PR TITLE
Add an extension point to OAuthAuthzRequest in OAuth2AuthzEndpoint

### PR DESCRIPTION
**Pull request by RivinduM**
Adding support to send PAR authorization requests without scope and response_type parameters.

As discussed in the mail[1], this PR adds an extension point to OAuthAuthzRequest which can be extended to add validations related to the PAR authorization flow.

[1] Clarification of spec compliance for Pushed Authorization Requests (PAR)

issue: https://github.com/wso2/product-is/issues/13416

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -387,15 +386,21 @@ private Response handleOAuthProblemException(OAuthMessage oAuthMessage, OAuthPro
             log.debug(e.getError(), e);
         }
 
+        String errorCode = e.getError() != null ? e.getError() : OAuth2ErrorCodes.INVALID_REQUEST;
+        String errorDescription = e.getDescription() != null ? e.getDescription() : e.getMessage();
+        String state = e.getState();
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:391]: since this code block is required only if the condition in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1827/files#diff-debebfb0b01b553afd77f73d24e1df466030d3d19b218e0d7b3005364a05939aR394 is not met, shall we move it after this condition?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 2]
```diff
@@ -1112,6 +1120,41 @@ public Map<String, RequestObjectBuilder> getRequestObjectBuilders() {
         return requestObjectBuilder;
     }
 
+    /**
+     * Returns an instance of OAuthAuthzRequest
+     *
+     * @return instance of OAuthAuthzRequest
+     */
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:1127]: incomplete method level comment

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 3]
```diff
@@ -2954,6 +2997,12 @@ private void parseOpenIDConnectConfig(OMElement oauthConfigElem) {
                     requestObjectEnabled = false;
                 }
             }
+            if (openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS(ConfigElements.
+                    OAUTH_AUTHZ_REQUEST)) != null) {
+                defaultOAuthAuthzRequestClassName =
+                        openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS(ConfigElements.
+                                OAUTH_AUTHZ_REQUEST)).getText().trim();
+            }
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:3005]:
```suggestion
                        OMElement oAuthAuthzRequest = openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS
                    (ConfigElements.OAUTH_AUTHZ_REQUEST));
            if (oAuthAuthzRequest != null) {
                defaultOAuthAuthzRequestClassName = oAuthAuthzRequest.getText().trim();
            }
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 4]
```diff
@@ -415,7 +415,12 @@ public static String getErrorPageURL(HttpServletRequest request, String errorCod
         if (request == null) {
             return redirectURL;
         }
-        return getRedirectURL(redirectURL, request);
+        if (isAllowAdditionalParamsFromErrorUrlEnabled()) {
+            // Appending additional parameters if the <AllowAdditionalParamsFromErrorUrl> config is enabled
+            return getRedirectURL(redirectURL, request);
+        } else {
+            return redirectURL;
+        }
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:423]: `
        String redirectURL = getErrorPageURL(errorCode, errorMessage, appName);
        if (request == null || !isAllowAdditionalParamsFromErrorUrlEnabled()) {
            return redirectURL;
        }
        return getRedirectURL(redirectURL, request);
`

Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:423]: suggestion for line 414 to 423:

```
        String redirectURL = getErrorPageURL(errorCode, errorMessage, appName);
        if (request == null || !isAllowAdditionalParamsFromErrorUrlEnabled()) {
            return redirectURL;
        }
        // Appending additional parameters if the <AllowAdditionalParamsFromErrorUrl> config is enabled
        return getRedirectURL(redirectURL, request);
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 5]
```diff
@@ -739,7 +744,7 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
                     // When request_uri requests come without redirect_uri, we need to append it to the SPQueryParams
                     // to be used in storing consent data
                     entry.setQueryString(entry.getQueryString() +
-                            "&" + PROP_REDIRECT_URI + "=" + params.getRedirectURI());
+                            "&" + PROP_REDIRECT_URI + "=" + URLEncoder.encode(params.getRedirectURI(), UTF_8));
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:747]: why is this required?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 1 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with e0a7995beb19d81a39eff11df56bc838eac1b687

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 2 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:1130]: Fixed with e0a7995beb19d81a39eff11df56bc838eac1b687

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 3 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Suggested change added with e0a7995beb19d81a39eff11df56bc838eac1b687

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 4 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:417]: Suggested change added with e0a7995beb19d81a39eff11df56bc838eac1b687

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 6]
```diff
@@ -2954,6 +2966,11 @@ private void parseOpenIDConnectConfig(OMElement oauthConfigElem) {
                     requestObjectEnabled = false;
                 }
             }
+            OMElement oAuthAuthzRequest = openIDConnectConfigElem.getFirstChildWithName(getQNameWithIdentityNS
+                    (ConfigElements.OAUTH_AUTHZ_REQUEST));
+            if (oAuthAuthzRequest != null) {
+                defaultOAuthAuthzRequestClassName = oAuthAuthzRequest.getText().trim();
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:2972]: The variable name is a bit misleading as it is returning the configured and not the default.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 7]
```diff
@@ -3223,6 +3240,7 @@ private class ConfigElements {
         public static final String SUPPORTED_CLAIMS = "OpenIDConnectClaims";
         public static final String REQUEST_OBJECT = "RequestObject";
         public static final String REQUEST_OBJECT_VALIDATOR = "RequestObjectValidator";
+        public static final String OAUTH_AUTHZ_REQUEST = "OAuthAuthzRequest";
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java:3243]:
```suggestion
        public static final String OAUTH_AUTHZ_REQUEST_CLASS = "OAuthAuthzRequestClass";
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 8]
```diff
@@ -1501,7 +1513,11 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
             setSPAttributeToRequest(oAuthMessage.getRequest(), validationResponse.getApplicationName(), tenantDomain);
         }
 
-        OAuthAuthzRequest oauthRequest = new CarbonOAuthAuthzRequest(oAuthMessage.getRequest());
+        String oauthAuthzRequestClassName = OAuthServerConfiguration.getInstance().getOAuthAuthzRequestClass();
+
+        OAuthAuthzRequest oauthRequest = defaultOAuthAuthzRequestClassName.equals(oauthAuthzRequestClassName) ?
+                new CarbonOAuthAuthzRequest(oAuthMessage.getRequest()) :
+                getOAuthAuthzRequest(oauthAuthzRequestClassName, oAuthMessage.getRequest());
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1520]: Lets move the logic of checking if the authz request class is default to a separate method and use it inside the getOAuthAuthzRequest method.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 9]
```diff
@@ -216,6 +218,9 @@ public class OAuth2AuthzEndpoint {
 
     private static final String OIDC_DIALECT = "http://wso2.org/oidc/claim";
 
+    private static final String defaultOAuthAuthzRequestClassName =
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:221]: Since the default class is defined in the OAuthServerConfiguration class let's get it from there. No need to define here.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 8 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1520]: The class resolving logic can also be in the `getOAuthAuthzRequest` method

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 10]
```diff
@@ -1541,6 +1557,45 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
...
+     * @throws OAuthProblemException thrown when initializing the OAuthAuthzRequestClass instance.
+     * @throws OAuthSystemException thrown when initializing the OAuthAuthzRequestClass instance.
+     */
+    private OAuthAuthzRequest getOAuthAuthzRequest(String oauthAuthzRequestClassName, HttpServletRequest request)
+            throws OAuthProblemException, OAuthSystemException {
+
+        OAuthAuthzRequest oAuthAuthzRequest;
+        try {
+            // Validations will be performed when initializing the class instance.
+            Class clazz = Thread.currentThread().getContextClassLoader().loadClass(oauthAuthzRequestClassName);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1576]: better to pass the Class itself as the param to avoid the loading at each request,

Also you may be able to restrict the class to be of this type by defining the param type as,
Class<? extends OAuthAuthzRequest>

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java (hunk 6 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Fixed with 79bab73a362a2021f071c540d7aed7a385ff3817

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 8 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with 79bab73a362a2021f071c540d7aed7a385ff3817

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 9 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with 79bab73a362a2021f071c540d7aed7a385ff3817

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 10 above)
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: Fixed with 79bab73a362a2021f071c540d7aed7a385ff3817

**Review by janakamarasena — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 11]
```diff
@@ -1541,6 +1552,79 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
         }
     }
 
+    /**
+     * Method to check whether the configured OAuthAuthzRequestImplementation is the default implementation.
+     *
+     * @return boolean whether the default class name is configured.
+     */
+    private boolean isDeafultOAuthAuthzRequestClassConfigured() {
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:1560]: typo

Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: fixed

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 12]
```diff
@@ -205,6 +205,9 @@ public class OAuthServerConfiguration {
             "org.wso2.carbon.identity.openidconnect.RequestObjectValidatorImpl";
     private String defaultCibaRequestValidatorClassName =
             "org.wso2.carbon.identity.openidconnect.CIBARequestObjectValidatorImpl";
+    private String oAuthAuthzRequestClassName;
+    public static final String DEFAULT_OAUTH_AUTHZ_REQUEST_CLASSNAME =
+            "org.wso2.carbon.identity.oauth2.model.CarbonOAuthAuthzRequest";
```
[the remark this answers is not in the document]
Comment (RivinduM) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: fixed

**Review by pulasthi7 — APPROVED** [review]
